src/jit/compiler.cpp

```diff
@@ -516,64 +487,450 @@ var_types    Compiler::argOrReturnTypeForStruct(unsigned size, CORINFO_CLASS_HAN
+                *wbPassStruct = SPK_ByValue;
+                useType = TYP_UNKNOWN;
+
+#else  //  _TARGET_XXX_
+
+                noway_assert(!"Unhandled TARGET in getArgTypeForStruct (with FEATURE_MULTIREG_ARGS=1)");
+
+#endif  //  _TARGET_XXX_
+
             }
         }
-#endif // FEATURE_MULTIREG_RET
+        else  // (structSize > MAX_PASS_MULTIREG_BYTES)
+        {
+            // We have a (large) struct that can't be replaced with a "primitive" type
+            // and can't be passed in multiple registers
 
-#if FEATURE_MULTIREG_ARGS
-        if (!forReturn)
+#if defined(_TARGET_X86_) || defined(_TARGET_ARM_)
+
+            // Otherwise we pass this struct by value on the stack 
+            // setup wbPassType and useType indicate that this is passed by value according to the X86/ARM32 ABI
+            *wbPassStruct = SPK_ByValue;
+            useType = TYP_UNKNOWN;
+
+#elif defined(_TARGET_AMD64_) || defined(_TARGET_ARM64_)
+
+            // Otherwise we pass this struct by reference to a copy
+            // setup wbPassType and useType indicate that this is passed using one register (by reference to a copy)
+            *wbPassStruct = SPK_ByReference;
+            useType = TYP_UNKNOWN;
+
+#else  //  _TARGET_XXX_
+
+            noway_assert(!"Unhandled TARGET in getArgTypeForStruct");
+
+#endif  //  _TARGET_XXX_
+
+        }
+    }
+
+    assert(*wbPassStruct != SPK_Unknown);  // This must get set to one of the valid values before we return
+    return useType;
+}
+
+//-----------------------------------------------------------------------------
+// getReturnTypeForStruct: 
+//     Get the type that is used to return values of the given struct type.
+//     If you already have retrieved the struct size then it should be passed
+//     as the third optional argument as it allows us to avoid an extra call
+//     to getClassSize(clsHnd)
+//
+// Arguments:
+//    clsHnd         - the handle for the struct type
+//    wbReturnStruct - information about how the struct is returned
+//    structSize     - the size of the struct type, 
+//                     or 0 if we should to call getClassSize(clsHnd)
+//
+// Return Value:
+//    When *wbReturnStruct is SPK_PrimitiveType this method's return value:
+//       is the primitive type used to return the struct.
+//    When *wbReturnStruct is SPK_ByReference this method's return value:
+//       is TYP_UNKNOWN when the struct type is returned using a return buffer
+//    When *wbReturnStruct is SPK_ByValue or SPK_ByValueAsHfa the return value:
+//       is TYP_STRUCT when the struct type is returned using multiple registers.
+//
+// Assumptions:
+//    The size must be the size of the given type.
+//    The given class handle must be for a value type (struct).
+//
+// Notes:
+//    About returning TYP_STRUCT:
+//        Whenever the return type is TYP_STRUCT it always means multiple registers
```

Comment:
I'm failing to resolve your distinction with immediately surrounding comments.  For example:
- Just above: "When \*wbPassStruct is SPK_ByValue or SPK_ByValueAsHfa the ***return value:***..."
- Just below: "If there are two or more elements in the HFA type then the ***return value*** is TYP_STRUCT...
But perhaps it's just me.  If you are convinced you wrote what you meant to write, then disregard.

Also, coding convention specifies "Return Value:" not "Return Type:", though I'm not sure how that's relevant.